// @instructure/ui-buttons/lib/CloseButton/props.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propTypes = exports.allowedProps = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _emotion = require("@instructure/emotion");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
const propTypes = {
  screenReaderLabel: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node]).isRequired,
  color: _propTypes.default.oneOf(['primary', 'primary-inverse']),
  interaction: _propTypes.default.oneOf(['enabled', 'disabled', 'readonly']),
  elementRef: _propTypes.default.func,
  size: _propTypes.default.oneOf(['small', 'medium', 'large']),
  onClick: _propTypes.default.func,
  margin: _emotion.ThemeablePropTypes.spacing,
  placement: _propTypes.default.oneOf(['start', 'end', 'static']),
  offset: _propTypes.default.oneOf(['none', 'x-small', 'small', 'medium']),
  type: _propTypes.default.oneOf(['button', 'submit', 'reset']),
  as: _propTypes.default.elementType,
  href: _propTypes.default.string,
  cursor: _propTypes.default.string,
  // Deprecated `string` tabIndex type
  // TODO: remove string type in v9
  tabIndex: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string])
};
exports.propTypes = propTypes;
const allowedProps = ['as', 'color', 'cursor', 'elementRef', 'href', 'interaction', 'margin', 'offset', 'onClick', 'placement', 'screenReaderLabel', 'size', 'tabIndex', 'type'];
exports.allowedProps = allowedProps;
