// @instructure/ui-a11y-utils/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FocusRegion", {
  enumerable: true,
  get: function () {
    return _FocusRegion.FocusRegion;
  }
});
Object.defineProperty(exports, "FocusRegionManager", {
  enumerable: true,
  get: function () {
    return _FocusRegionManager.FocusRegionManager;
  }
});
Object.defineProperty(exports, "KeyboardFocusRegion", {
  enumerable: true,
  get: function () {
    return _KeyboardFocusRegion.KeyboardFocusRegion;
  }
});
Object.defineProperty(exports, "ScreenReaderFocusRegion", {
  enumerable: true,
  get: function () {
    return _ScreenReaderFocusRegion.ScreenReaderFocusRegion;
  }
});
Object.defineProperty(exports, "hasVisibleChildren", {
  enumerable: true,
  get: function () {
    return _hasVisibleChildren.hasVisibleChildren;
  }
});
Object.defineProperty(exports, "scopeTab", {
  enumerable: true,
  get: function () {
    return _scopeTab.scopeTab;
  }
});

var _FocusRegion = require("./FocusRegion");

var _FocusRegionManager = require("./FocusRegionManager");

var _hasVisibleChildren = require("./hasVisibleChildren");

var _KeyboardFocusRegion = require("./KeyboardFocusRegion");

var _scopeTab = require("./scopeTab");

var _ScreenReaderFocusRegion = require("./ScreenReaderFocusRegion");
