// @instructure/ui-checkbox/lib/Checkbox/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Checkbox = void 0;
Object.defineProperty(exports, "CheckboxFacade", {
  enumerable: true,
  get: function () {
    return _CheckboxFacade.CheckboxFacade;
  }
});
Object.defineProperty(exports, "ToggleFacade", {
  enumerable: true,
  get: function () {
    return _ToggleFacade.ToggleFacade;
  }
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _keycode = _interopRequireDefault(require("keycode"));

var _FormFieldMessages = require("@instructure/ui-form-field/lib/FormFieldMessages");

var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");

var _console = require("@instructure/console");

var _isActiveElement = require("@instructure/ui-dom-utils/lib/isActiveElement.js");

var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");

var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");

var _View = require("@instructure/ui-view/lib/View");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _emotion = require("@instructure/emotion");

var _CheckboxFacade = require("./CheckboxFacade");

var _ToggleFacade = require("./ToggleFacade");

var _styles = _interopRequireDefault(require("./styles"));

var _props = require("./props");

var _dec, _dec2, _dec3, _class, _class2;

/**
---
category: components
tags: toggle, switch
---
@tsProps
**/
let Checkbox = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, null), _dec3 = (0, _testable.testable)(), _dec(_class = _dec2(_class = _dec3(_class = (_class2 = class Checkbox extends _react.Component {
  constructor(props) {
    super(props);
    this._defaultId = void 0;
    this._input = null;
    this.ref = null;

    this.handleRef = el => {
      this.ref = el;
    };

    this.handleChange = e => {
      const _this$props = this.props,
            onChange = _this$props.onChange,
            disabled = _this$props.disabled,
            checked = _this$props.checked,
            readOnly = _this$props.readOnly;

      if (disabled || readOnly) {
        e.preventDefault();
        return;
      }

      if (typeof checked === 'undefined') {
        this.setState({
          checked: !this.state.checked
        });
      }

      if (typeof onChange === 'function') {
        onChange(e);
      }
    };

    this.handleKeyDown = e => {
      if (this.props.variant === 'toggle' && e.keyCode === _keycode.default.codes.enter) {
        this._input && this._input.click();
        e.preventDefault();
      }
    };

    this.handleFocus = () => {
      this.setState({
        focused: true
      });
    };

    this.handleBlur = () => {
      this.setState({
        focused: false
      });
    };

    this.handleMouseOver = () => {
      this.setState({
        hovered: true
      });
    };

    this.handleMouseOut = () => {
      this.setState({
        hovered: false
      });
    };

    this.state = {
      focused: false,
      hovered: false,
      checked: typeof props.checked === 'undefined' ? !!props.defaultChecked : void 0
    };

    if (typeof props.checked === 'undefined') {
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'checked' does not exist on type 'Readonl... Remove this comment to see the full error message
      this.state.checked = !!props.defaultChecked;
    }

    this._defaultId = props.deterministicId();
  }

  componentDidMount() {
    var _this$props$makeStyle, _this$props2;

    if (this._input) {
      this._input.indeterminate = this.props.indeterminate;
    }

    (_this$props$makeStyle = (_this$props2 = this.props).makeStyles) === null || _this$props$makeStyle === void 0 ? void 0 : _this$props$makeStyle.call(_this$props2);
  }

  componentDidUpdate(prevProps) {
    var _this$props$makeStyle2, _this$props3;

    if (prevProps.indeterminate !== this.props.indeterminate && this._input) {
      this._input.indeterminate = this.props.indeterminate || false;
    }

    (_this$props$makeStyle2 = (_this$props3 = this.props).makeStyles) === null || _this$props$makeStyle2 === void 0 ? void 0 : _this$props$makeStyle2.call(_this$props3);
  }

  get id() {
    return this.props.id || this._defaultId;
  }

  get checked() {
    return typeof this.props.checked === 'undefined' ? this.state.checked : this.props.checked;
  }

  get focused() {
    return (0, _isActiveElement.isActiveElement)(this._input);
  }

  focus() {
    this._input && this._input.focus();
  }

  renderFacade() {
    const _this$props4 = this.props,
          size = _this$props4.size,
          disabled = _this$props4.disabled,
          variant = _this$props4.variant,
          label = _this$props4.label,
          readOnly = _this$props4.readOnly,
          indeterminate = _this$props4.indeterminate,
          labelPlacement = _this$props4.labelPlacement,
          themeOverride = _this$props4.themeOverride;
    const _this$state = this.state,
          hovered = _this$state.hovered,
          focused = _this$state.focused;
    (0, _console.logError)(!(variant === 'simple' && labelPlacement !== 'end'), `[Checkbox] The \`simple\` variant does not support the \`labelPlacement\` property.  Use the \`toggle\` variant instead.`);

    if (variant === 'toggle') {
      return (0, _emotion.jsx)(_ToggleFacade.ToggleFacade, {
        disabled: disabled,
        size: size,
        focused: focused,
        checked: this.checked,
        readOnly: readOnly,
        labelPlacement: labelPlacement,
        themeOverride: themeOverride
      }, label);
    } else {
      return (0, _emotion.jsx)(_CheckboxFacade.CheckboxFacade, {
        size: size,
        hovered: hovered,
        focused: focused,
        checked: this.checked,
        indeterminate: indeterminate,
        themeOverride: themeOverride
      }, label);
    }
  }

  renderMessages() {
    const messages = this.props.messages;
    return messages && messages.length > 0 ? (0, _emotion.jsx)(_View.View, {
      display: "block",
      margin: "small 0 0"
    }, (0, _emotion.jsx)(_FormFieldMessages.FormFieldMessages, {
      messages: messages
    })) : null;
  }

  render() {
    const _this$props5 = this.props,
          disabled = _this$props5.disabled,
          readOnly = _this$props5.readOnly,
          value = _this$props5.value,
          onKeyDown = _this$props5.onKeyDown,
          onFocus = _this$props5.onFocus,
          onBlur = _this$props5.onBlur,
          onMouseOver = _this$props5.onMouseOver,
          onMouseOut = _this$props5.onMouseOut,
          indeterminate = _this$props5.indeterminate,
          variant = _this$props5.variant,
          styles = _this$props5.styles;
    const props = (0, _omitProps.omitProps)(this.props, Checkbox.allowedProps);
    (0, _console.logError)(!(variant === 'toggle' && indeterminate), `[Checkbox] The \`toggle\` variant does not support the \`indeterminate\` property. Use the \`simple\` variant instead.`);
    /* eslint-disable jsx-a11y/mouse-events-have-key-events */

    return (0, _emotion.jsx)("div", {
      css: styles === null || styles === void 0 ? void 0 : styles.checkbox,
      onMouseOver: (0, _createChainedFunction.createChainedFunction)(onMouseOver, this.handleMouseOver),
      onMouseOut: (0, _createChainedFunction.createChainedFunction)(onMouseOut, this.handleMouseOut),
      ref: this.handleRef
    }, (0, _emotion.jsx)("input", Object.assign({}, props, {
      id: this.id,
      value: value,
      type: "checkbox",
      ref: c => {
        this._input = c;
      },
      disabled: disabled || readOnly,
      "aria-checked": indeterminate ? 'mixed' : void 0,
      css: styles === null || styles === void 0 ? void 0 : styles.input,
      onChange: this.handleChange,
      onKeyDown: (0, _createChainedFunction.createChainedFunction)(onKeyDown, this.handleKeyDown),
      onFocus: (0, _createChainedFunction.createChainedFunction)(onFocus, this.handleFocus),
      onBlur: (0, _createChainedFunction.createChainedFunction)(onBlur, this.handleBlur),
      checked: this.checked
    })), (0, _emotion.jsx)("label", {
      htmlFor: this.id,
      css: styles === null || styles === void 0 ? void 0 : styles.control
    }, this.renderFacade(), this.renderMessages()));
    /* eslint-enable jsx-a11y/mouse-events-have-key-events */
  }

}, _class2.displayName = "Checkbox", _class2.componentId = 'Checkbox', _class2.propTypes = _props.propTypes, _class2.allowedProps = _props.allowedProps, _class2.defaultProps = {
  size: 'medium',
  variant: 'simple',
  disabled: false,
  inline: false,
  indeterminate: false,
  readOnly: false,
  labelPlacement: 'end'
}, _class2)) || _class) || _class) || _class);
exports.Checkbox = Checkbox;
var _default = Checkbox;
exports.default = _default;
