// @instructure/ui-buttons/es/ToggleButton/props.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import PropTypes from 'prop-types';
import { PositionPropTypes } from '@instructure/ui-position';
const propTypes = {
  screenReaderLabel: PropTypes.string.isRequired,
  renderTooltipContent: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
  renderIcon: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
  status: PropTypes.oneOf(['pressed', 'unpressed']).isRequired,
  as: PropTypes.elementType,
  interaction: PropTypes.oneOf(['enabled', 'disabled', 'readonly']),
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  elementRef: PropTypes.func,
  onClick: PropTypes.func,
  color: PropTypes.oneOf(['primary', 'primary-inverse', 'secondary', 'success', 'danger']),
  isShowingTooltip: PropTypes.bool,
  mountNode: PositionPropTypes.mountNode,
  placement: PositionPropTypes.placement,
  constrain: PositionPropTypes.constrain
};
const allowedProps = ['as', 'color', 'constrain', 'elementRef', 'interaction', 'isShowingTooltip', 'mountNode', 'onClick', 'placement', 'renderIcon', 'renderTooltipContent', 'screenReaderLabel', 'size', 'status'];
export { propTypes, allowedProps };
