// @instructure/ui-checkbox/lib/Checkbox/CheckboxFacade/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.CheckboxFacade = void 0;

var _react = require("react");

var _SVGIcon2 = require("@instructure/ui-svg-images/lib/SVGIcon");

var _IconCheckMarkSolid2 = require("@instructure/ui-icons/lib/IconCheckMarkSolid.js");

var _emotion = require("@instructure/emotion");

var _styles = _interopRequireDefault(require("./styles"));

var _theme = _interopRequireDefault(require("./theme"));

var _props = require("./props");

var _dec, _class, _class2, _SVGIcon, _IconCheckMarkSolid;

/**
---
parent: Checkbox
---
@tsProps
**/
let CheckboxFacade = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = (_class2 = class CheckboxFacade extends _react.Component {
  constructor() {
    super(...arguments);
    this.ref = null;

    this.handleRef = el => {
      this.ref = el;
    };
  }

  componentDidMount() {
    var _this$props$makeStyle, _this$props;

    (_this$props$makeStyle = (_this$props = this.props).makeStyles) === null || _this$props$makeStyle === void 0 ? void 0 : _this$props$makeStyle.call(_this$props);
  }

  componentDidUpdate() {
    var _this$props$makeStyle2, _this$props2;

    (_this$props$makeStyle2 = (_this$props2 = this.props).makeStyles) === null || _this$props$makeStyle2 === void 0 ? void 0 : _this$props$makeStyle2.call(_this$props2);
  }

  renderIcon() {
    if (this.props.indeterminate) {
      return _SVGIcon || (_SVGIcon = (0, _emotion.jsx)(_SVGIcon2.SVGIcon, {
        viewBox: "0 0 1920 1920",
        inline: false
      }, (0, _emotion.jsx)("rect", {
        x: "140",
        y: "820",
        width: "1640",
        height: "280"
      })));
    } else if (this.props.checked) {
      return _IconCheckMarkSolid || (_IconCheckMarkSolid = (0, _emotion.jsx)(_IconCheckMarkSolid2.IconCheckMarkSolid, {
        inline: false
      }));
    } else {
      return null;
    }
  }

  render() {
    const _this$props3 = this.props,
          children = _this$props3.children,
          styles = _this$props3.styles;
    return (0, _emotion.jsx)("span", {
      css: styles === null || styles === void 0 ? void 0 : styles.checkboxFacade,
      ref: this.handleRef
    }, (0, _emotion.jsx)("span", {
      css: styles === null || styles === void 0 ? void 0 : styles.facade,
      "aria-hidden": "true"
    }, this.renderIcon()), (0, _emotion.jsx)("span", {
      css: styles === null || styles === void 0 ? void 0 : styles.label
    }, children));
  }

}, _class2.displayName = "CheckboxFacade", _class2.componentId = 'CheckboxFacade', _class2.propTypes = _props.propTypes, _class2.allowedProps = _props.allowedProps, _class2.defaultProps = {
  checked: false,
  focused: false,
  hovered: false,
  size: 'medium',
  indeterminate: false
}, _class2)) || _class);
exports.CheckboxFacade = CheckboxFacade;
var _default = CheckboxFacade;
exports.default = _default;
