// @instructure/ui-a11y-content/lib/PresentationContent/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.PresentationContent = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");

var _props = require("./props");

const _excluded = ["children"];

/**
---
category: components/utilities
---
@module PresentationContent
@tsProps
**/
class PresentationContent extends _react.Component {
  constructor() {
    super(...arguments);
    this.ref = null;

    this.handleRef = el => {
      const elementRef = this.props.elementRef;
      this.ref = el;

      if (typeof elementRef === 'function') {
        elementRef(el);
      }
    };
  }

  render() {
    const _this$props = this.props,
          children = _this$props.children,
          props = (0, _objectWithoutProperties2.default)(_this$props, _excluded);
    const ElementType = (0, _getElementType.getElementType)(PresentationContent, this.props);
    return /*#__PURE__*/_react.default.createElement(ElementType // @ts-expect-error TODO: `ref` prop causes: "Expression produces a union type that is too complex to represent.ts(2590)"
    , Object.assign({}, (0, _passthroughProps.passthroughProps)(props), {
      "aria-hidden": "true",
      ref: this.handleRef
    }), children);
  }

}

exports.PresentationContent = PresentationContent;
PresentationContent.displayName = "PresentationContent";
PresentationContent.propTypes = _props.propTypes;
PresentationContent.allowedProps = _props.allowedProps;
PresentationContent.defaultProps = {
  as: 'span',
  children: null
};
var _default = PresentationContent;
exports.default = _default;
