// @instructure/ui-a11y-content/lib/AccessibleContent/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.AccessibleContent = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");

var _PresentationContent = require("../PresentationContent");

var _ScreenReaderContent = require("../ScreenReaderContent");

var _props = require("./props");

const _excluded = ["alt", "children"];

/**
---
category: components/utilities
---
@module AccessibleContent
@tsProps
**/
class AccessibleContent extends _react.Component {
  constructor() {
    super(...arguments);
    this.ref = null;

    this.handleRef = el => {
      this.ref = el;
    };
  }

  render() {
    const _this$props = this.props,
          alt = _this$props.alt,
          children = _this$props.children,
          props = (0, _objectWithoutProperties2.default)(_this$props, _excluded); // any cast is needed to prevent Expression produces a union type that is too complex to represent errors

    const ElementType = (0, _getElementType.getElementType)(AccessibleContent, this.props);
    return /*#__PURE__*/_react.default.createElement(ElementType, Object.assign({}, (0, _passthroughProps.passthroughProps)(props), {
      ref: this.handleRef
    }), /*#__PURE__*/_react.default.createElement(_ScreenReaderContent.ScreenReaderContent, null, alt), /*#__PURE__*/_react.default.createElement(_PresentationContent.PresentationContent, null, children));
  }

}

exports.AccessibleContent = AccessibleContent;
AccessibleContent.displayName = "AccessibleContent";
AccessibleContent.propTypes = _props.propTypes;
AccessibleContent.allowedProps = _props.allowedProps;
AccessibleContent.defaultProps = {
  as: 'span',
  children: null
};
var _default = AccessibleContent;
exports.default = _default;
