// @instructure/ui-buttons/lib/BaseButton/props.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propTypes = exports.allowedProps = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _emotion = require("@instructure/emotion");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
const propTypes = {
  children: _propTypes.default.node,
  type: _propTypes.default.oneOf(['button', 'submit', 'reset']),
  size: _propTypes.default.oneOf(['small', 'medium', 'large']),
  elementRef: _propTypes.default.func,
  as: _propTypes.default.elementType,
  interaction: _propTypes.default.oneOf(['enabled', 'disabled', 'readonly']),
  color: _propTypes.default.oneOf(['primary', 'primary-inverse', 'secondary', 'success', 'danger']),
  focusColor: _propTypes.default.oneOf(['info', 'inverse']),
  display: _propTypes.default.oneOf(['inline-block', 'block']),
  textAlign: _propTypes.default.oneOf(['start', 'center']),
  shape: _propTypes.default.oneOf(['rectangle', 'circle']),
  withBackground: _propTypes.default.bool,
  withBorder: _propTypes.default.bool,
  isCondensed: _propTypes.default.bool,
  margin: _emotion.ThemeablePropTypes.spacing,
  cursor: _propTypes.default.string,
  href: _propTypes.default.string,
  onClick: _propTypes.default.func,
  onKeyDown: _propTypes.default.func,
  renderIcon: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),
  // Deprecated `string` tabIndex type
  // TODO: remove string type in v9
  tabIndex: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string])
};
exports.propTypes = propTypes;
const allowedProps = ['as', 'children', 'color', 'cursor', 'display', 'elementRef', 'focusColor', 'href', 'interaction', 'isCondensed', 'margin', 'onClick', 'onKeyDown', 'renderIcon', 'shape', 'size', 'tabIndex', 'textAlign', 'type', 'withBackground', 'withBorder'];
exports.allowedProps = allowedProps;
