// @instructure/ui-portal/lib/Portal/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Portal = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");

var _textDirectionContextConsumer = require("@instructure/ui-i18n/lib/textDirectionContextConsumer.js");

var _canUseDOM = require("@instructure/ui-dom-utils/lib/canUseDOM.js");

var _console = require("@instructure/console");

var _props = require("./props");

const _excluded = ["open", "insertAt", "onOpen", "onClose", "mountNode", "children", "elementRef", "dir"];

var _dec, _class, _class2;

/**
---
category: components/utilities
---
@module Portal
@tsProps
**/
let Portal = (_dec = (0, _textDirectionContextConsumer.textDirectionContextConsumer)(), _dec(_class = (_class2 = class Portal extends _react.Component {
  constructor(props) {
    super(props);
    this.DOMNode = null;
    this.ref = null;

    this.handleRef = el => {
      const elementRef = this.props.elementRef;
      this.ref = el;
      this.DOMNode = el;

      if (typeof elementRef === 'function') {
        elementRef(el);
      }
    };

    if (!_canUseDOM.canUseDOM) {
      return;
    }

    this.state = {
      mountNode: this.findMountNode(props)
    };
  }

  componentDidMount() {
    if (!_canUseDOM.canUseDOM) {
      return;
    } // If Portal is mounting in an open condition fire onOpen handler


    if (this.props.open && typeof this.props.onOpen === 'function') {
      this.props.onOpen(this.DOMNode);
    }
  }

  componentDidUpdate(prevProps) {
    if (!_canUseDOM.canUseDOM) {
      return;
    }

    const mountNode = this.findMountNode(this.props);

    if (mountNode !== this.state.mountNode) {
      // set state here to make the component re-render
      this.setState({
        mountNode
      });
    } // If Portal was closed but is now open fire onOpen handler


    if (this.props.open && !prevProps.open && typeof this.props.onOpen === 'function') {
      this.props.onOpen(this.DOMNode);
    } // If Portal was open but is now closed fire onClose handler


    if (!this.props.open && prevProps.open && typeof this.props.onClose === 'function') {
      this.props.onClose();
    }
  }

  componentWillUnmount() {
    if (!_canUseDOM.canUseDOM) {
      return;
    }

    this.removeNode(); // If Portal was open fire onClose handler

    if (this.props.open && typeof this.props.onClose === 'function') {
      this.props.onClose();
    }
  }

  removeNode() {
    if (this.DOMNode && this.DOMNode.parentNode && typeof this.DOMNode.parentNode.removeChild === 'function') {
      this.DOMNode.parentNode.removeChild(this.DOMNode);
      this.handleRef(null);
    }
  }

  insertNode() {
    const _this$props = this.props,
          open = _this$props.open,
          insertAt = _this$props.insertAt,
          onOpen = _this$props.onOpen,
          onClose = _this$props.onClose,
          mountNode = _this$props.mountNode,
          children = _this$props.children,
          elementRef = _this$props.elementRef,
          dir = _this$props.dir,
          props = (0, _objectWithoutProperties2.default)(_this$props, _excluded); // Create node if it doesn't already exist

    if (!this.DOMNode) {
      const node = document.createElement('span');
      const attributes = { ...(0, _passthroughProps.passthroughProps)(props),
        dir
      };
      Object.keys(attributes).forEach(name => {
        node.setAttribute(name, attributes[name]);
      });
      this.handleRef(node);
    } // Append node to container if it isn't already


    if (this.DOMNode && this.DOMNode.parentNode !== this.state.mountNode) {
      if (insertAt === 'bottom') {
        this.state.mountNode.appendChild(this.DOMNode);
      } else {
        this.state.mountNode.insertBefore(this.DOMNode, this.state.mountNode.firstChild);
      }
    }

    return this.DOMNode;
  }

  findMountNode(props) {
    let mountNode;

    if (typeof props.mountNode === 'function') {
      mountNode = props.mountNode();
    } else if (props.mountNode) {
      mountNode = props.mountNode;
    }

    if (!mountNode || !mountNode.nodeName) {
      mountNode = document.body;
    }

    return mountNode;
  }

  get node() {
    return this.DOMNode;
  }

  render() {
    if (!_canUseDOM.canUseDOM) {
      (0, _console.warn)(false, `It looks like you are trying to render <Portal> on the server side, which is not supported!
         Please be aware that this will result in a no-op!
         Read more about it here: https://instructure.design/#server-side-rendering`);
      return null;
    }

    const children = this.props.children;
    return this.props.open && _react.default.Children.count(children) > 0 ? /*#__PURE__*/_reactDom.default.createPortal(children, this.insertNode()) : null;
  }

}, _class2.displayName = "Portal", _class2.propTypes = _props.propTypes, _class2.allowedProps = _props.allowedProps, _class2.defaultProps = {
  open: false,
  insertAt: 'bottom',
  onOpen: () => {},
  onClose: () => {},
  mountNode: null,
  children: null,
  elementRef: () => {}
}, _class2)) || _class);
exports.Portal = Portal;
var _default = Portal;
exports.default = _default;
