// @instructure/ui-color-utils/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "alpha", {
  enumerable: true,
  get: function () {
    return _alpha.alpha;
  }
});
Object.defineProperty(exports, "color2hex", {
  enumerable: true,
  get: function () {
    return _conversions.color2hex;
  }
});
Object.defineProperty(exports, "colorToHex8", {
  enumerable: true,
  get: function () {
    return _conversions.colorToHex8;
  }
});
Object.defineProperty(exports, "colorToHsla", {
  enumerable: true,
  get: function () {
    return _conversions.colorToHsla;
  }
});
Object.defineProperty(exports, "colorToHsva", {
  enumerable: true,
  get: function () {
    return _conversions.colorToHsva;
  }
});
Object.defineProperty(exports, "colorToRGB", {
  enumerable: true,
  get: function () {
    return _conversions.colorToRGB;
  }
});
Object.defineProperty(exports, "contrast", {
  enumerable: true,
  get: function () {
    return _contrast.contrast;
  }
});
Object.defineProperty(exports, "darken", {
  enumerable: true,
  get: function () {
    return _darken.darken;
  }
});
Object.defineProperty(exports, "isValid", {
  enumerable: true,
  get: function () {
    return _isValid.isValid;
  }
});
Object.defineProperty(exports, "lighten", {
  enumerable: true,
  get: function () {
    return _lighten.lighten;
  }
});

var _alpha = require("./alpha");

var _darken = require("./darken");

var _lighten = require("./lighten");

var _contrast = require("./contrast");

var _isValid = require("./isValid");

var _conversions = require("./conversions");
