// @instructure/ui-buttons/es/CloseButton/props.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import PropTypes from 'prop-types';
import { ThemeablePropTypes } from '@instructure/emotion';
const propTypes = {
  screenReaderLabel: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  color: PropTypes.oneOf(['primary', 'primary-inverse']),
  interaction: PropTypes.oneOf(['enabled', 'disabled', 'readonly']),
  elementRef: PropTypes.func,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  onClick: PropTypes.func,
  margin: ThemeablePropTypes.spacing,
  placement: PropTypes.oneOf(['start', 'end', 'static']),
  offset: PropTypes.oneOf(['none', 'x-small', 'small', 'medium']),
  type: PropTypes.oneOf(['button', 'submit', 'reset']),
  as: PropTypes.elementType,
  href: PropTypes.string,
  cursor: PropTypes.string,
  // Deprecated `string` tabIndex type
  // TODO: remove string type in v9
  tabIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
};
const allowedProps = ['as', 'color', 'cursor', 'elementRef', 'href', 'interaction', 'margin', 'offset', 'onClick', 'placement', 'screenReaderLabel', 'size', 'tabIndex', 'type'];
export { propTypes, allowedProps };
