// @instructure/ui-position/lib/Position/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Position = void 0;

var _react = require("react");

var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");

var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");

var _ensureSingleChild = require("@instructure/ui-react-utils/lib/ensureSingleChild.js");

var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");

var _addPositionChangeListener = require("@instructure/ui-dom-utils/lib/addPositionChangeListener.js");

var _findDOMNode = require("@instructure/ui-dom-utils/lib/findDOMNode.js");

var _shallowEqual = require("@instructure/ui-utils/lib/shallowEqual.js");

var _deepEqual = require("@instructure/ui-utils/lib/deepEqual.js");

var _debounce = require("@instructure/debounce");

var _testable = require("@instructure/ui-testable/lib/testable.js");

var _Portal = require("@instructure/ui-portal/lib/Portal");

var _emotion = require("@instructure/emotion");

var _styles = _interopRequireDefault(require("./styles"));

var _theme = _interopRequireDefault(require("./theme"));

var _props = require("./props");

var _calculateElementPosition = require("../calculateElementPosition");

var _dec, _dec2, _dec3, _class, _class2;

/**
---
category: components/utilities
---
@tsProps
**/
let Position = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec3 = (0, _testable.testable)(), _dec(_class = _dec2(_class = _dec3(_class = (_class2 = class Position extends _react.Component {
  constructor(props) {
    super(props);
    this.ref = null;
    this._id = void 0;
    this._timeouts = [];
    this._listener = null;
    this._content = void 0;
    this._target = void 0;

    this.handleRef = el => {
      const elementRef = this.props.elementRef;
      this.ref = el;

      if (typeof elementRef === 'function') {
        elementRef(el);
      }
    };

    this.handlePortalOpen = () => {
      this.position();

      if (this.props.shouldTrackPosition) {
        this.startTracking();
      }

      this._timeouts.push(setTimeout(() => {
        if (this.state.positioned && typeof this.props.onPositioned === 'function') {
          this.props.onPositioned({
            top: this.state.style.top,
            left: this.state.style.left,
            placement: this.state.placement
          });
        }
      }, 0));
    };

    this.position = () => {
      this.setState({
        positioned: true,
        ...this.calculatePosition(this.props)
      });
    };

    this.state = {
      positioned: false,
      ...this.calculatePosition(props)
    };
    this.position = (0, _debounce.debounce)(this.position, 0, {
      leading: false,
      trailing: true
    });
    this._id = this.props.id || props.deterministicId();
  }

  shouldComponentUpdate(nextProps, nextState, nextContext) {
    return !(0, _deepEqual.deepEqual)(this.state, nextState) || !(0, _shallowEqual.shallowEqual)(this.props, nextProps) || !(0, _shallowEqual.shallowEqual)(this.context, nextContext);
  }

  componentDidMount() {
    var _this$props$makeStyle, _this$props;

    this.toggleLocatorAttributes(true);
    (_this$props$makeStyle = (_this$props = this.props).makeStyles) === null || _this$props$makeStyle === void 0 ? void 0 : _this$props$makeStyle.call(_this$props);
  }

  componentDidUpdate(prevProps, prevState) {
    var _this$props$makeStyle2, _this$props2;

    this.position();
    this.toggleLocatorAttributes(true);

    if (this.props.shouldTrackPosition !== prevProps.shouldTrackPosition) {
      this.props.shouldTrackPosition ? this.startTracking() : this.stopTracking();
    }

    const _this$state = this.state,
          style = _this$state.style,
          placement = _this$state.placement;

    if (style && prevState.style && (placement !== prevState.placement || style.top !== prevState.style.top || style.left !== prevState.style.left) && typeof this.props.onPositionChanged === 'function') {
      this.props.onPositionChanged({
        top: style.top,
        left: style.left,
        placement
      });
    }

    (_this$props$makeStyle2 = (_this$props2 = this.props).makeStyles) === null || _this$props$makeStyle2 === void 0 ? void 0 : _this$props$makeStyle2.call(_this$props2);
  }

  componentWillUnmount() {
    ;
    this.position.cancel();
    this.stopTracking();

    this._timeouts.forEach(timeout => clearTimeout(timeout));

    this.toggleLocatorAttributes(false);
  }

  toggleLocatorAttributes(set) {
    // We have to find the actual DOM nodes and append the attributes
    // directly, as we can't be sure when safe cloning the child that
    // it will accept the data attribute as a prop
    this.toggleLocatorAttribute((0, _findDOMNode.findDOMNode)(this._content), Position.contentLocatorAttribute, set);
    this.toggleLocatorAttribute((0, _findDOMNode.findDOMNode)(this._target), Position.targetLocatorAttribute, set);
  }

  toggleLocatorAttribute(node, locator, set) {
    if (node && node.hasAttribute) {
      if (set && !node.hasAttribute(locator)) {
        ;
        node.setAttribute(locator, this._id);
      }

      if (!set && node.hasAttribute(locator)) {
        ;
        node.removeAttribute(locator);
      }
    }
  }

  calculatePosition(props) {
    return (0, _calculateElementPosition.calculateElementPosition)(this._content, this._target, {
      placement: props.placement,
      offsetX: props.offsetX,
      offsetY: props.offsetY,
      constrain: props.constrain,
      container: props.mountNode,
      over: props.shouldPositionOverTarget
    });
  }

  startTracking() {
    this._listener = this._listener || (0, _addPositionChangeListener.addPositionChangeListener)(this._target, this.position);
  }

  stopTracking() {
    if (this._listener) {
      this._listener.remove();

      this._listener = null;
    }
  } // content that needs to be positioned relative to the target


  renderContent() {
    let content = (0, _ensureSingleChild.ensureSingleChild)(this.props.children);

    if (content) {
      var _this$props$styles;

      content = (0, _safeCloneElement.safeCloneElement)(content, {
        ref: el => {
          this._content = el;
        },
        style: {
          boxSizing: 'border-box',
          zIndex: (_this$props$styles = this.props.styles) === null || _this$props$styles === void 0 ? void 0 : _this$props$styles.zIndex,
          ...content.props.style,
          ...this.state.style
        },
        ...(content.props.className && {
          className: content.props.className
        }),
        [Position.contentLocatorAttribute]: this._id
      });
      content = (0, _emotion.jsx)(_Portal.Portal, {
        open: true,
        onOpen: this.handlePortalOpen,
        mountNode: this.props.mountNode,
        insertAt: this.props.insertAt
      }, content);
    }

    return content;
  }

  renderTarget() {
    const target = (0, _callRenderProp.callRenderProp)(this.props.renderTarget);

    if (target) {
      return (0, _safeCloneElement.safeCloneElement)(target, {
        ref: el => {
          this._target = el;
        },
        [Position.targetLocatorAttribute]: this._id
      });
    } else if (this.props.target) {
      this._target = (0, _callRenderProp.callRenderProp)(this.props.target);
    }

    return null;
  }

  render() {
    const props = {
      [Position.locatorAttribute]: this._id
    };
    return (0, _emotion.jsx)("span", Object.assign({}, props, {
      ref: this.handleRef
    }), this.renderTarget(), this.renderContent());
  }

}, _class2.displayName = "Position", _class2.componentId = 'Position', _class2.allowedProps = _props.allowedProps, _class2.propTypes = _props.propTypes, _class2.defaultProps = {
  placement: 'bottom center',
  mountNode: null,
  insertAt: 'bottom',
  constrain: 'window',
  offsetX: 0,
  offsetY: 0,
  shouldTrackPosition: true,
  shouldPositionOverTarget: false,
  children: null
}, _class2.locatorAttribute = 'data-position', _class2.targetLocatorAttribute = 'data-position-target', _class2.contentLocatorAttribute = 'data-position-content', _class2)) || _class) || _class) || _class);
exports.Position = Position;
var _default = Position;
exports.default = _default;
