// @instructure/ui-tooltip/lib/Tooltip/props.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propTypes = exports.allowedProps = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _PositionPropTypes = require("@instructure/ui-position/lib/PositionPropTypes.js");

var _element = require("@instructure/ui-prop-types/lib/element.js");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
const propTypes = {
  elementRef: _propTypes.default.func,
  children: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]).isRequired,
  renderTip: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]).isRequired,
  isShowingContent: _propTypes.default.bool,
  defaultIsShowingContent: _propTypes.default.bool,
  as: _propTypes.default.elementType,
  // eslint-disable-line react/require-default-props
  on: _propTypes.default.oneOfType([_propTypes.default.oneOf(['click', 'hover', 'focus']), _propTypes.default.arrayOf(_propTypes.default.oneOf(['click', 'hover', 'focus']))]),
  color: _propTypes.default.oneOf(['primary', 'primary-inverse']),
  placement: _PositionPropTypes.PositionPropTypes.placement,
  mountNode: _PositionPropTypes.PositionPropTypes.mountNode,
  constrain: _PositionPropTypes.PositionPropTypes.constrain,
  offsetX: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  offsetY: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  positionTarget: _propTypes.default.oneOfType([_element.element, _propTypes.default.func]),
  onShowContent: _propTypes.default.func,
  onHideContent: _propTypes.default.func
};
exports.propTypes = propTypes;
const allowedProps = ['elementRef', 'children', 'renderTip', 'isShowingContent', 'defaultIsShowingContent', 'as', 'on', 'color', 'placement', 'mountNode', 'constrain', 'offsetX', 'offsetY', 'positionTarget', 'onShowContent', 'onHideContent'];
exports.allowedProps = allowedProps;
