// @instructure/ui-menu/lib/Menu/MenuLocator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "MenuItemGroupLocator", {
  enumerable: true,
  get: function () {
    return _MenuItemGroupLocator.MenuItemGroupLocator;
  }
});
Object.defineProperty(exports, "MenuItemLocator", {
  enumerable: true,
  get: function () {
    return _MenuItemLocator.MenuItemLocator;
  }
});
exports.MenuLocator = void 0;

var _locator = require("@instructure/ui-test-locator/lib/utils/locator.js");

var _uiTestQueries = require("@instructure/ui-test-queries");

var _PopoverLocator = require("@instructure/ui-popover/lib/Popover/PopoverLocator");

var _index = require("./index");

var _MenuItemLocator = require("./MenuItem/MenuItemLocator");

var _MenuItemGroupLocator = require("./MenuItemGroup/MenuItemGroupLocator");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
// @ts-expect-error ts-migrate(7016) FIXME: Could not find a declaration file for module '@ins... Remove this comment to see the full error message
// eslint-disable-next-line no-restricted-imports
const customMethods = {
  findAllItems: function () {
    return _MenuItemLocator.MenuItemLocator.findAll(...arguments);
  },
  findItem: function () {
    return _MenuItemLocator.MenuItemLocator.find(...arguments);
  },
  findAllGroups: function () {
    return _MenuItemGroupLocator.MenuItemGroupLocator.findAll(...arguments);
  },
  findGroup: function () {
    return _MenuItemGroupLocator.MenuItemGroupLocator.find(...arguments);
  },
  findPopoverTrigger: function () {
    return _PopoverLocator.PopoverLocator.findTrigger(...arguments);
  },
  findPopoverContent: function () {
    const _parseQueryArguments = (0, _uiTestQueries.parseQueryArguments)(...arguments),
          element = _parseQueryArguments.element,
          selector = _parseQueryArguments.selector,
          options = _parseQueryArguments.options;

    return _PopoverLocator.PopoverLocator.findContent(element, selector, { ...options,
      customMethods: { ...options.customMethods,
        ...customMethods
      }
    });
  }
}; // @ts-expect-error ts-migrate(2339) FIXME: Property 'selector' does not exist on type 'typeof... Remove this comment to see the full error message

const MenuLocator = (0, _locator.locator)(_index.Menu.selector, customMethods);
exports.MenuLocator = MenuLocator;
