// @instructure/ui-menu/lib/Menu/props.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propTypes = exports.allowedProps = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _PositionPropTypes = require("@instructure/ui-position/lib/PositionPropTypes.js");

var _controllable = require("@instructure/ui-prop-types/lib/controllable.js");

var _Children = require("@instructure/ui-prop-types/lib/Children.js");

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
const propTypes = {
  children: _Children.Children.oneOf(['MenuItem', 'MenuItemGroup', 'MenuItemSeparator', 'Menu']),
  label: _propTypes.default.string,
  disabled: _propTypes.default.bool,
  trigger: _propTypes.default.node,
  placement: _PositionPropTypes.PositionPropTypes.placement,
  defaultShow: _propTypes.default.bool,
  show: (0, _controllable.controllable)(_propTypes.default.bool, 'onToggle', 'defaultShow'),
  onToggle: _propTypes.default.func,
  onSelect: _propTypes.default.func,
  onDismiss: _propTypes.default.func,
  onFocus: _propTypes.default.func,
  onMouseOver: _propTypes.default.func,
  onKeyDown: _propTypes.default.func,
  onKeyUp: _propTypes.default.func,
  menuRef: _propTypes.default.func,
  popoverRef: _propTypes.default.func,
  mountNode: _PositionPropTypes.PositionPropTypes.mountNode,
  constrain: _PositionPropTypes.PositionPropTypes.constrain,
  shouldHideOnSelect: _propTypes.default.bool,
  shouldFocusTriggerOnClose: _propTypes.default.bool,
  type: _propTypes.default.oneOf(['flyout']),
  id: _propTypes.default.string,
  withArrow: _propTypes.default.bool,
  offsetX: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  offsetY: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number])
};
exports.propTypes = propTypes;
const allowedProps = ['children', 'label', 'disabled', 'trigger', 'placement', 'defaultShow', 'show', 'onToggle', 'onSelect', 'onDismiss', 'onFocus', 'onMouseOver', 'onKeyDown', 'onKeyUp', 'menuRef', 'popoverRef', 'mountNode', 'constrain', 'shouldHideOnSelect', 'shouldFocusTriggerOnClose', 'type', 'id', 'withArrow', 'offsetX', 'offsetY'];
exports.allowedProps = allowedProps;
