// @instructure/console/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "assert", {
  enumerable: true,
  get: function () {
    return _console2.assert;
  }
});
Object.defineProperty(exports, "consoleLog", {
  enumerable: true,
  get: function () {
    return _console.log;
  }
});
Object.defineProperty(exports, "debug", {
  enumerable: true,
  get: function () {
    return _console2.debug;
  }
});
Object.defineProperty(exports, "error", {
  enumerable: true,
  get: function () {
    return _console2.error;
  }
});
Object.defineProperty(exports, "info", {
  enumerable: true,
  get: function () {
    return _console2.info;
  }
});
Object.defineProperty(exports, "log", {
  enumerable: true,
  get: function () {
    return _console2.log;
  }
});
Object.defineProperty(exports, "logAssert", {
  enumerable: true,
  get: function () {
    return _console.assert;
  }
});
Object.defineProperty(exports, "logDebug", {
  enumerable: true,
  get: function () {
    return _console.debug;
  }
});
Object.defineProperty(exports, "logError", {
  enumerable: true,
  get: function () {
    return _console.error;
  }
});
Object.defineProperty(exports, "logInfo", {
  enumerable: true,
  get: function () {
    return _console.info;
  }
});
Object.defineProperty(exports, "logWarn", {
  enumerable: true,
  get: function () {
    return _console.warn;
  }
});
Object.defineProperty(exports, "logWarnDeprecated", {
  enumerable: true,
  get: function () {
    return _console.warnDeprecated;
  }
});
Object.defineProperty(exports, "warn", {
  enumerable: true,
  get: function () {
    return _console2.warn;
  }
});
Object.defineProperty(exports, "warnDeprecated", {
  enumerable: true,
  get: function () {
    return _console2.warnDeprecated;
  }
});

var _console = require("@instructure/console");

var _console2 = require("./console");
